pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,144 @@
+	var leCtx context.Context
+	var leCancel context.CancelFunc
+	for {
+		select {
+		case <-ctx.Done():
+			le.logger.Info("Shutting down LeaderElection...")
+			if leCancel != nil {
+				leCancel()
+			}
+			return nil
+		case <-time.After(le.Config.ReelectionPeriod.Duration):
+			isLeader, err := le.IsLeader(ctx)
+			if err != nil {
+				le.logger.Errorf("Failed to elect the backup-restore leader: %v", err)
+
+				// set the CurrentState of backup-restore.
+				// stops the Running Snapshotter.
+				// wait for Reelection to happen.
+				le.CurrentState = StateUnknown
+				le.logger.Infof("backup-restore is in: %v", le.CurrentState)
+				if le.Callbacks.OnStoppedLeading != nil && leCtx != nil {
+					leCancel()
+					le.Callbacks.OnStoppedLeading()
+				}
+				le.logger.Info("waiting for Re-election...")
+				continue
+			}
+
+			if isLeader && (le.CurrentState == StateFollower || le.CurrentState == StateUnknown || le.CurrentState == StateCandidate) {
+				// backup-restore becomes the Leader backup-restore.
+				// set the CurrentState of backup-restore.
+				// update the snapshotter object with latest snapshotter object.
+				// start the snapshotter.
+				le.CurrentState = StateLeader
+				le.logger.Infof("backup-restore became: %v", le.CurrentState)
+
+				if le.Callbacks.OnStartedLeading != nil {
+					leCtx, leCancel = context.WithCancel(ctx)
+					le.logger.Info("backup-restore started leading...")
+					le.Callbacks.OnStartedLeading(leCtx)
+				}
+			} else if isLeader && le.CurrentState == StateLeader {
+				le.logger.Debug("no change in leadershipStatus...")
+			} else if !isLeader && le.CurrentState == StateLeader {
+				// backup-restore lost the election and becomes Follower.
+				// set the CurrentState of backup-restore.
+				// stop the Running snapshotter.
+				le.CurrentState = StateFollower
+				le.logger.Info("backup-restore lost the election")
+				le.logger.Infof("backup-restore became: %v", le.CurrentState)
+
+				if le.Callbacks.OnStoppedLeading != nil && leCtx != nil {
+					leCancel()
+					le.Callbacks.OnStoppedLeading()
+				}
+			} else if !isLeader && le.CurrentState == StateUnknown {
+				le.CurrentState = StateFollower
+				le.logger.Infof("backup-restore changed the state from %v to %v", StateUnknown, le.CurrentState)
+			} else if !isLeader && le.CurrentState == StateFollower {
+				le.logger.Infof("backup-restore currentState: %v", le.CurrentState)
+			}
+		}
+	}
+}
+
+// IsLeader checks whether the current backup-restore is leader or not.
+func (le *LeaderElector) IsLeader(ctx context.Context) (bool, error) {
+	le.logger.Info("checking the leadershipStatus...")
+	var endPoint string
+	client, err := etcdutil.GetTLSClientForEtcd(le.EtcdConnectionConfig)
+	if err != nil {
+		return false, &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to create etcd client: %v", err),
+		}
+	}
+	defer client.Close()
+
+	if len(le.EtcdConnectionConfig.Endpoints) > 0 {
+		endPoint = le.EtcdConnectionConfig.Endpoints[0]
```

Comment:
It is the endpoint of `etcd` which comes fom the [etcdConnectionConfig](https://github.com/gardener/etcd-backup-restore/blob/5dfcc1f848a9f325d41a24eae4defb70d997c215/pkg/etcdutil/types.go#L35-L49), I think it should be passed through environment variable but I'm also not so sure, so this thing need to be take care.

cc @abdasgupta @shreyas-s-rao @timuthy @aaronfern

Reply:
yeah, endpoints of etcd comes as an array but actually Question was "How to make sure that endpoints present at 0th index in a array is always endpoint of their corresponding etcd?" @amshuman-kr also raised this question.
answer to this questions could be that etcd endpoints should passed as environment variable to their corresponding sidecar.

Reply:
Passing different config to different sidecars will be hard given that we have only one `StatefulSet` which creates all the pods of the ETCD cluster.

The member sidecars need to still distinguish between themselves and this can be achieved by passing the pod name (which would also be the ETCD member name according to the proposal) as an environment variable using `fieldRef` mechanism.

The code here should ideally either iterate through the member names and select the correct endpoint of the member by comparing with the pod/member name passed via env variable or it should rely on some convention in configuration where the endpoint can be computed based on the passed pod/member name.